torch/distributed/_tensor/placement_types.py

```diff
@@ -378,11 +378,20 @@ class DTensorSpec:
 
     def __hash__(self) -> int:
         # hashing and equality check for DTensorSpec are used to cache the sharding
-        # propagation results. We only need to consider the mesh, placements and shape
+        # propagation results. We only need to consider the mesh, placements, shape
+        # dtype and stride.
         # Caveat: we need to keep this in mind and sync hash and eq if we add more
-        # fields to them,
+        # fields to them.
         if self.tensor_meta is not None:
-            return hash((self.mesh, self.placements, self.tensor_meta.shape))
+            return hash(
+                (
+                    self.mesh,
+                    self.placements,
+                    self.tensor_meta.shape,
```

Comment:
I guess we don't need dtype, shall we delete `dtype` in both hash and eq then?

Reply:
Do we expect DTensor not to be used in mixed precision scenario? because this can cause wrong result in the mixed precision situation.

Reply:
for mixed precision we still have the same sharding propagation rules, so dtype won't affect how we conduct sharding propagations IMO, so likely we don't need dtype here

Reply:
but the output sharding contains wrong dtype right?

Reply:
Correct me if I was wrong:
it seems the usage of prop cache becomes quite restrictive: only all attributes match then is there a cache hit. Probably we adjust the cached OutputSharding object's dtype before returning it?